### PR TITLE
Merging upstream

### PR DESCRIPTION
**Pull request by kohanis**
Merging up to v2.3.1.1 of upstream + some fixes

### CODE REVIEW

**Review by ManlyMarco — COMMENTED** [review]
It seems fine but b70734949454005c72c0bd72f2cec83fd269461e is making it hard to review. Ideally I'd prefer to not have that commit in this PR but in a separate one.

Edit: How does this relate to #79? I'm planning to merge that one once I get to test it with bep6 nightlies.

**Comment by kohanis** [comment]
Fair, removed it for now. As for 79, I don't see any incompatibilities, but better check later. I was actually planning to do a full merge to 2.3-prerelease.3 later, but I was finding some regressions in my tests with 2.3, so haven't yet
I'd prefer to move this to draft for now, as it's something to think about. There's a reason why original harmony doesn't distribute standard versions (or rather now they do, but only ref ones), might have to do the same here. Because of net5+ directives. It will probably need to be split into net5.0, netcoreapp3.1 and net45

**Comment by kohanis** [comment]
Now it should be okay. Used same method as in harmony. Main targets are now net35;net45;netcoreapp3.0;net5.0, and netstandard2.0 is now ref only

**Review by ManlyMarco — DISMISSED** [review]

**Comment by ManlyMarco** [comment]
I merged the previous PR, looks like it created some conflicts. I'll merge this one as well if you can resolve them.

**Comment by kohanis** [comment]
I'll wait a bit for discussion on 79

**Comment by ManlyMarco** [comment]
@kohanis Everything's merged now so it should be fine to revisit this.

**Comment by kohanis** [comment]
Just noting that I currently have everything synchronised with harmony master, but I'm waiting for 2.3, which I understand should be out soon https://github.com/pardeike/Harmony/commit/ff82ff4f85bb077c632fae6aa5e8c81b7b03aa7f

**Comment by kohanis** [comment]
Well, it's basically done, maybe just add github workflows later on

Binary incompatibilities(kind of):
`HarmonyLib.InlineSignature` removed (moved to internal in harmony, not used at all in harmonyx)
`HarmonyLib.MethodType.Async` / `HarmonyLib.AccessTools.AsyncMoveNext` hidden from net35 build

Comments:
`ValueTuple` shims removed - they are now in `MonoMod.Backports`
`__args` now unpacks itself back as in harmony
netstandard2.0 build is now ref only

**Comment by ManlyMarco** [comment]
On a glance it seems fine, but looks like this will need some testing in different heavily modded games before a merge.

> ValueTuple shims removed - they are now in MonoMod.Backports

Is there a chance those shims were used by plugins, which will now break?

> Also, harmony now embeds pdb into the release dll(for thin version, but harmonyx doesn't have fat version anyway). Should it be here?

Embedding debug symbols in release is fine, yes.

**Comment by kohanis** [comment]
> Is there a chance those shims were used by plugins, which will now break?

Oh, those were internal, I missed it. So, no

> Embedding debug symbols in release is fine, yes.

Well, it's not embedded in 2.3.1 now) Will merge up to 2.3.1.1 soon, only changes for project files, not sources

**Comment by kohanis** [comment]
@Meivyn you can try this. Here is different approach to `GetExecutingAssembly`, and one another `master` fix. Might help. Based on the log, I doubt it, but maybe.

**Comment by Meivyn** [comment]
That does indeed not fix the issue. I can't tell if your replacement works how it's intended to because I'm not sure to understand what that's even fixing.

**Comment by kohanis** [comment]
I believe last push should fix #107
And @Meivyn, if your reverse patch target is extern(and I think it is, there was a regression in #79), that should help too

Yes, it should be calli and not a generated shenanigans, but there are currently some issues with managed calli with `DynamicMethodDefinition`.

**Comment by Meivyn** [comment]
The target method is not an extern method, it's a random method I choose to test the reverse patch:

image

I also confirm that your latest push fixes the issue with the native methods patches.

**Comment by kohanis** [comment]
It looks like inlining. Try calling through a delegate to check, instead of directly

**Comment by Meivyn** [comment]
That indeed did the trick. Though why is that required now? It was working fine in the previous version of Harmony.
That didn't solve the exception when patching though.

**Comment by CptMoore** [comment]
Looks like @kohanis fixes to native detour patcher indeed fixes issue #107, thanks.
Could you also use
```<MonoModRuntimeDetour>25.1.0</MonoModRuntimeDetour>```
instead of the prelease version, as it includes a vital unity+linux fix.

**Comment by kohanis** [comment]
> That indeed did the trick. Though why is that required now? It was working fine in the previous version of Harmony (and it breaks existing patches).

Frankly I don't see any reason why mono wouldn't inline a purely throwing method without `[MethodImpl(MethodImplOptions.NoInlining)]`.
Is the usage exactly the same? If method with reverse patch call is called at least once before patching itself, this behaviour is not surprising

**Comment by Meivyn** [comment]
Watch out, bumping the MonoMod version requires Harmony to target net452 instead of net45, otherwise it won't use the correct version of Mono.Cecil.

> Is the usage exactly the same? If method with reverse patch call is called at least once before patching itself, this behaviour is not surprising

Yes, the usage is the same, as I said it was breaking patches that did work before, and it breaks my test one as well, and as you can see that method isn't called before patching.

**Comment by kohanis** [comment]
> Yes, the usage is the same, as I said it was breaking patches that did work before and it breaks my test one as well, and as you can see that method isn't called before patching. Patching happens in `Init` which is always called before `OnEnable` as it's the constructor.

Cannot confirm it. Used latest BSIPA 4.3.2 with your sample on PlateUp without `MethodImplOptions.NoInlining`, got `NotImplementedException`

**Comment by Meivyn** [comment]
Yeah, I actually cannot reproduce the same behavior with this patch as well. It seems to happen only the

**Comment by kohanis** [comment]
Okay, reverse patches are broken in general in 1.11.0+, affirmative. GC is collecting them. I'll see what I can come up with

**Comment by kohanis** [comment]
Far from perfect, but let's call it a hotfix

**Comment by Meivyn** [comment]
That doesn't fix the inline issue for my test patch, however it seems to have fixed the broken patch I mentioned earlier. Guess I'm only missing a fix for the IL error that happens when reverse patching some methods before using this PR in BSIPA.

**Comment by Meivyn** [comment]
Maybe not. Some patches seem to still be broken but aren't throwing any errors. I wouldn't be able to tell which either...

**Comment by kohanis** [comment]
Restructured. Last push in this PR, unless fixes related to it are needed. I'm planning some structural changes of project based on this PR, but separately

**Comment by kohanis** [comment]
Pardon, that fix was important. Also, friendly reminder that without `HarmonyX.Ref` package project will not be built as is, and CI is also not adapted for this

**Comment by kohanis** [comment]
As far as I can tell it should work the same way, but with two packages. Not sure about integration with jenkins though

**Comment by CptMoore** [comment]
whats missing for this to be merged? CI? HarmonyX.Ref?

**Review by ManlyMarco — COMMENTED** [review]
Sorry it took me so long to review, most of the changes are just syntax fluff which makes it hard to spot and review any material changes.

As far as I can see this is good to merge and shouldn't break binary backwards compatibility. There may be some minor source code breaks but that's fine.
It might require a major version bump still, but that's for later.

My main worry is workflows and dealing with the ref/full situation. Ideally builds and packages would stay in the same format as they are now, in my opinion at least.

If no one else raises any issues with this PR in the meantime I'll merge this later next week.

File: .github/workflows/publish_release.yml
Comment (kohanis) [reply .github/workflows/publish_release.yml:29]: There is no point in building with old SDKs, and old SDKs won't work with new targets anyway

**Comment by kohanis** [comment]
Theoretically I can merge netstandard references into main package, will check.

File: .github/workflows/publish_release.yml
Comment (CptMoore) [reply .github/workflows/publish_release.yml:29]: I also only build with newest dotnet while targeting net472. That strategy worked fine for a couple of years now

**Comment by kohanis** [comment]
Okay, there's a [reason for separate package (harmony discord)](https://discord.com/channels/131466550938042369/674571535570305060/1213858710351519764). 
But, after rethinking, I realized that the restriction on netstandard does not apply to harmonyx. Harmony's problem with it is ilmerge.
So, this should work. I plan to later squish this into the first commit

**Review by ManlyMarco — APPROVED** [review]

**Pull request merged**
